### PR TITLE
trilinos: simplify some variants

### PR DESCRIPTION
**Pull request by sethrj**
- Use hwloc iff kokkos
    See #19119, also the HWLOC tpl name was misspelled so this was being ignored before.
- Remove package meta-options
    - XSDK settings and "all opt packages" are not used anywhere
    - all optional packages are dangerous
- Update dependencies, removing unused
    - GLM, X11 are unused (x11 lacks dependency specs too)
    - Python variant is more like a TPL so rearrange that
    - Gtest internal package shouldn't be compiled or exported
    - Add MPI4PY requirement for pytrilinos
- Group top-level dependencies
- Automatically enable teuchos and remove variant
    - Basically everything in trilinos needs teuchos
    - Simplify dealii dependencies (@jppelteret @luca-heltai)
- Remove metis variant, enable if zoltan or if other transitive dependencies use it
  - Trilinos packages usually conflict if the metis/parmetis options are not locked to the availability of strumpack and superlu-dist 
- Rename basker variant
    - The Basker solver is part of amesos2 but is clearer without the extra prefix
scoping.
- Default to disabling external trilinos dependencies (except for mpi)
  - Some of these are troublemakers (e.g. mumps) and get turned off by most downstream spack packages

### CODE REVIEW

**Comment by kuberry** [comment]
Hi @sethrj, what should ~hwloc become under the current changes?

**Comment by sethrj** [comment]
@kuberry The hwloc TPL is only used by kokkos and shylu (as optional dependencies in both places), and it's already an implicit dependency when building `+mpi`, so I see little benefit to keeping it as an explicit variant. (So disabling kokkos will disable hwloc.) Are there known compatibility issues with `kokkos` using `hwloc`? Should it default to "off"? The fewer variants `trilinos` has the more stable the build will be.

**Comment by kuberry** [comment]
I’m not aware of the specifics of the hwloc/kokkos usage combo with regard to variants. Thanks for getting a head start on updating the xyce recipe. It still contains an ~hwloc. Would you mind removing it as well? I was just asking the previous question from the perspective of changing it to be compatible with removing the hwloc variant.

**Comment by kuberry** [comment]
Just tried building Trilinos and got the following error (among other hwloc errors):
/var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos-12.12.1-4qwqv5j7dvc5iomuyqzbft3lbqmstfwc/spack-src/packages/kokkos/core/src/impl/Kokkos_hwloc.cpp:334:61: error: no member named 'allowed_cpuset' in 'hwloc_obj'

In the xyce recipe, I just removed the ~hwloc constraint.

**Comment by sethrj** [comment]
Thanks @kuberry , I missed that. I just successfully built with `+kokkos+hwloc`, so maybe I'll enable hwloc only for trilinos 13+? That should solve your problem since xyce is hardwired to the older version.

**Comment by kuberry** [comment]
>> spack install trilinos
```==> Error: More than one spec in string: hypre cuda_arch10```

This is caused by changes in nalu-wind.

**Comment by kuberry** [comment]
There is now a conflict by default with ``` spack install trilinos``` as netcdf is not built by default (I think this is the right way to go), however exodus is turned on by default and requires netcdf. Can we disable exodus by default?

**Comment by kuberry** [comment]
It looks like `hypre cuda_arch` in nalu-wind needs an equal sign added, i.e. `hypre cuda_arch=`

**Comment by sethrj** [comment]
Thanks again @kuberry . I verified that all of the downstream trilinos dependencies concretize (with clingo at least) after I reverted #20623 which seems to confuse `nalu-wind`.

**Comment by kuberry** [comment]
I can now concretize again as well. Still having hwloc issues with Trilinos 12.12.1 called from Xyce. I'm looking into it.

**Comment by sethrj** [comment]
> I can now concretize again as well. Still having hwloc issues with Trilinos 12.12.1 called from Xyce. I'm looking into it.

Whoops, I only updated the variant requirement; trilinos is still explicitly configuring it (and it's available because of the upstream dependencies). I'll fix.

**Comment by kuberry** [comment]
Just tested and this works for the Xyce recipe. Would you mind applying the following patch for the Xyce recipe?:

patch.txt
```
index 77de6591cc..f30d200144 100644
--- a/var/spack/repos/builtin/packages/xyce/package.py
+++ b/var/spack/repos/builtin/packages/xyce/package.py
@@ -50,13 +50,25 @@ class Xyce(CMakePackage):
     depends_on('trilinos~mpi', when='~mpi')
     depends_on('trilinos+mpi', when='+mpi')
 
+    # CUDA options
+    depends_on('trilinos~cuda', when='~cuda')
+    depends_on('trilinos+cuda', when='+cuda')
+
+    # debug options
+    depends_on('trilinos~debug', when='~debug')
+    depends_on('trilinos+debug', when='+debug')
+
+    # shared options
+    depends_on('trilinos~shared', when='~shared')
+    depends_on('trilinos+shared', when='+shared')
+
+    # wrapper options
+    depends_on('trilinos~wrapper', when='~wrapper')
+    depends_on('trilinos+wrapper', when='+wrapper')
+
     # The default settings for various Trilinos variants would require the
     # installation of many more packages than are needed for Xyce.
-    # The default variants in Trilinos have been set for several
-    # applications, namely xSDK, deal.ii, and DTK. Future changes to the
-    # Trilinos recipe will disable all packages by default. At that
-    # point, these ~variants can be removed from the following recipes.
-    depends_on('trilinos~adios2~boost~cgns~chaco~cuda~cuda_rdc~debug~dtk~exodus~float~hypre~ifpack2~intrepid~intrepid2~ipo~matio~mesquite~metis~minitensor~ml~muelu~mumps~netcdf~openmp~phalanx~piro~pnetcdf~python~rol~rythmos~shards~shared~shylu~stk~stratimikos~strumpack~superlu~superlu-dist~teko~tempus~wrapper~zlib~zoltan2')
+    depends_on('trilinos~float~ifpack2~ml~muelu~zoltan2')
 
     def cmake_args(self):
         spec = self.spec
```

Then apply with `git apply patch.txt`

As you can see from these changes, this significantly simplifies to Trilinos call from the Xyce recipe (Thanks!)

**Comment by sethrj** [comment]
@kuberry I simplified the variant propagation to
```diff
-    # MPI options must be consistent with Trilinos
-    depends_on('trilinos~mpi', when='~mpi')
-    depends_on('trilinos+mpi', when='+mpi')
+    # Propagate variants to trilinos:
+    for _variant in ('mpi', 'cuda', 'debug', 'shared', 'wrapper'):
+        depends_on('trilinos~' + _variant, when='~' + _variant)
+        depends_on('trilinos+' + _variant, when='+' + _variant)
```
however, clingo concretization correctly fails with the error:
```
==> Error: variant "cuda" not found in package "xyce" [required from package "xyce"]
```

I also notice a couple of other potential issues in the xyce recipe:
- your `cmake_args` checks for `'+shared'` which will silently be "false" when the variant doesn't exist. Use `self.define_from_variant('BUILD_SHARED_LIBS', 'shared')` to simplify the logic with additional error checking too.
- you define a variant `'build_type'` that overrides the default cmake build type (which has a default of 'relwithdebinfo'). I would recommend sticking with the default spack one for consistency, unless debug symbols cause xyce to fail to build.

**Comment by kuberry** [comment]
New error:
```
/var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilinos
             -12.12.1-g43oco3cxuf6girdwoj7lyutq5voewoq/spack-src/packages/zoltan/src/tpls/third_librar
             y.c:111:10: error: implicit declaration of function 'Zoltan_ParMetis' is invalid in C99 [
             -Werror,-Wimplicit-function-declaration]
     1211        rc = Zoltan_ParMetis(zz, part_sizes, num_imp, imp_gids, imp_lids,
     1212             ^
     1213    1 error generated.
  >> 1214    make[2]: *** [packages/zoltan/src/CMakeFiles/zoltan.dir/tpls/third_library.c.o] Error 1
     1215    make[2]: *** Waiting for unfinished jobs....
```
Seems metis related (which was just removed)

**Comment by kuberry** [comment]
Previously,
```
        options.append(define_tpl_enable('METIS'))
        options.append(define_tpl_enable('ParMETIS', 'metis'))
```
would have set metis and ParMETIS to False by setting them false. Now, this is no longer being set to False explicitly, so Trilinos thinks that it is available (because it may be a default in Cmake).

**Comment by kuberry** [comment]
Seems to be building right now with Trilinos 13+, so this may be a Trilinos 12 issue.

**Comment by keitat** [comment]
> Seems to be building right now with Trilinos 13+, so this may be a Trilinos 12 issue.

Let me talk to Trilinos (Zoltan?) folks.

**Comment by sethrj** [comment]
> > Seems to be building right now with Trilinos 13+, so this may be a Trilinos 12 issue.
> 
> Let me talk to Trilinos (Zoltan?) folks.

I looked at it, and it might just an untested combination of `TPL_ENABLE_METIS=ON TPL_ENABLE_ParMETIS=OFF`. I could just force parmetis to be on if zoltan is on.

This is again the downside of infinitely configurable options: the known-working combinations become diminishingly small.

**Comment by sethrj** [comment]
Good news is that the e4s pipeline now passes!

@kuberry to get xyce working I could force `^metis ^partmetis` dependencies on trilinos when `trilinos@:12.18 +zoltan`?

**Comment by kuberry** [comment]
Hi @sethrj, would you mind providing the source code that you would use to force ^metis ^partmetis dependencies on trilinos when trilinos@:12.18 +zoltan?

**Comment by sethrj** [comment]
@kuberry I'm thinking the following:
```
$ git diff
diff --git a/var/spack/repos/builtin/packages/trilinos/package.py b/var/spack/repos/builtin/packages/trilin
index 2cb58f639e..62880022bb 100644
--- a/var/spack/repos/builtin/packages/trilinos/package.py
+++ b/var/spack/repos/builtin/packages/trilinos/package.py
@@ -191,9 +191,9 @@ class Trilinos(CMakePackage, CudaPackage):
             description='Compile with Tpetra')
     variant('trilinoscouplings', default=False,
             description='Compile with TrilinosCouplings')
-    variant('zoltan',       default=True,
+    variant('zoltan',       default=False,
             description='Compile with Zoltan')
-    variant('zoltan2',      default=True,
+    variant('zoltan2',      default=False,
             description='Compile with Zoltan2')

     # Internal package options (alphabetical order)
@@ -370,6 +370,8 @@ class Trilinos(CMakePackage, CudaPackage):
     depends_on('netcdf-c+mpi', when="+netcdf~pnetcdf+mpi")
     depends_on('netcdf-c+mpi+parallel-netcdf', when="+netcdf+pnetcdf@master,12.12.1:")
     depends_on('parallel-netcdf', when="+netcdf+pnetcdf@master,12.12.1:")
+    depends_on('metis', when='@:12 +zoltan')
+    depends_on('parmetis', when='@:12 +zoltan')
     depends_on('parmetis', when='+scorec')
     depends_on('cgns', when='+cgns')
     depends_on('adios2', when='+adios2')
```

with this change, `xyce` concretizes to include both metis and parmetis so the zoltan build error should be solved.

**Comment by kuberry** [comment]
This does fix the issue for building Trilinos 12.12.1 as Xyce calls it. However, Xyce's recipe calls `~ml`. If we try to build with ml, there is a different error (we don't, but I'm sure someone else could).

Unfortunately, it seems a bit cryptic:
```  >> 3643    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilin
             os-12.12.1-urezoqcqxo4ddmxm6j3236ac6ziagfos/spack-src/packages/ml/src/Coarsen/ml_aggreg
             ate.c:2097:3: error: implicit declaration of function 'ML_Operator_Copy_Statistics' is 
             invalid in C99 [-Werror,-Wimplicit-function-declaration]
     3644      ML_Operator_Copy_Statistics(mat,permuted_Amat);
     3645      ^
     3646    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilin
             os-12.12.1-urezoqcqxo4ddmxm6j3236ac6ziagfos/spack-src/packages/ml/src/Coarsen/ml_aggreg
             ate.c:2097:3: note: did you mean 'ML_Operator_ReportStatistics'?
     3647    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilin
             os-12.12.1-urezoqcqxo4ddmxm6j3236ac6ziagfos/spack-src/packages/ml/src/Operator/ml_op_ut
             ils.h:110:13: note: 'ML_Operator_ReportStatistics' declared here
     3648    extern void ML_Operator_ReportStatistics(ML_Operator *mat, char *appendlabel,
     3649                ^
  >> 3650    /var/folders/zk/1zr9bdg53s3_d7qxmv2m5f80002l2h/T/pakuber/spack-stage/spack-stage-trilin
             os-12.12.1-urezoqcqxo4ddmxm6j3236ac6ziagfos/spack-src/packages/ml/src/Coarsen/ml_aggreg
             ate.c:2416:5: error: implicit declaration of function 'ML_Operator_Copy_Statistics' is 
             invalid in C99 [-Werror,-Wimplicit-function-declaration]
     3651        ML_Operator_Copy_Statistics(Pmat,newP);
     3652        ^
     3653    2 errors generated.
  >> 3654    make[2]: *** [packages/ml/src/CMakeFiles/ml.dir/Coarsen/ml_aggregate.c.o] Error 1
```

**Comment by sethrj** [comment]
@kuberry is that with 12.x or with 13.0.1? ☹️

**Comment by kuberry** [comment]
12.x, 12.12.1 to be specific.

**Comment by sethrj** [comment]
> 12.x, 12.12.1 to be specific.

I hope that the number of users pinned to 12.x is pretty small; certainly the exascale peeps tend to be working off trilinos develop. I can add a conflict marker for this case. I suggest we try to focus on getting the minimal 12.x requirement from xyce working, and the trilinos combinations needed by E4S (which are passing on the CI). And I think those both are ok now, right?

> This does fix the issue for building Trilinos 12.12.1 as Xyce calls it. However, Xyce's recipe calls `~ml`. If we try to build with ml, there is a different error (we don't, but I'm sure someone else could).
> 
> Unfortunately, it seems a bit cryptic

In summary: we could be stuck in trilinos configure combinatorics hell for the rest of our lives if we go down this rabbit hole. Believe me, I've been there.

**Comment by sethrj** [comment]
Pings for maintainers whose package changed due to the trilinos updates:
- camellia: @CamelliaDPG
- datatransferkit: @Rombur
- dealii: @jppelteret  @luca-heltai
- nalu-wind: @jrood-nrel
- omega-h: @ibaned
- percept:
- phist: @jthies
- quinoa:
- xsdk: @balay @luszczek

**Review by ibaned — DISMISSED** [review]

**Review by kuberry — DISMISSED** [review]

**Comment by sethrj** [comment]
@eugeneswalker can you take a quick look and sign off on this? It's another step in the quest to make trilinos builds more reliable and intuitive. The gitlab CI pipeline is passing so that should mean it works with the latest E4S definition. Thanks!

**Review by jrood-nrel — DISMISSED** [review]

File: var/spack/repos/builtin/packages/xsdk/package.py [hunk 1]
```diff
@@ -64,17 +64,17 @@ class Xsdk(BundlePackage):
     depends_on('superlu-dist@5.2.2', when='@0.3.0')
     depends_on('superlu-dist@xsdk-0.2.0', when='@xsdk-0.2.0')
 
-    depends_on('trilinos@develop+hypre+superlu-dist+metis+hdf5~mumps+boost~suite-sparse+tpetra+nox+ifpack2+zoltan2+amesos2~exodus~dtk+intrepid2+shards+stratimikos gotype=int cxxstd=14',
+    depends_on('trilinos@develop+hypre+superlu-dist+hdf5~mumps+boost~suite-sparse+tpetra+nox+ifpack2+zoltan2+amesos2~exodus~dtk+intrepid2+shards+stratimikos gotype=int cxxstd=14',
```
Comment (balay) [inline var/spack/repos/builtin/packages/xsdk/package.py]: looks like `+zoltan` is an internal dependency for `+zoltan2`?
```
spack spec xsdk:

1. "+zoltan2" conflicts with "trilinos~zoltan"
```
I'm trying out the following fix 

https://github.com/spack/spack/commit/224581142602452d1cf8f6cedf7a019d13832f85

https://gitlab.com/xsdk-project/spack-xsdk/-/pipelines/337795803

Comment (sethrj) [reply var/spack/repos/builtin/packages/xsdk/package.py]: Ooooh, I think I misread the trilinos dependencies here. It's got a *test* required dependencies but not a *build* dep. I will revert this. Thanks!

**Comment by eugeneswalker** [comment]
> @eugeneswalker can you take a quick look and sign off on this? It's another step in the quest to make trilinos builds more reliable and intuitive. The gitlab CI pipeline is passing so that should mean it works with the latest E4S definition. Thanks!

Yes of course, and thank you for doing this. I do have to submit an additional PR so that we are building trilinos with all variants turned on as part of the official PR- and merge-to-develop pipelines on this repo. See this:
https://github.com/spack/spack/pull/24908

File: var/spack/repos/builtin/packages/xsdk/package.py (hunk 1 above)
Comment (sethrj) [reply var/spack/repos/builtin/packages/xsdk/package.py]: Wait sorry, Zoltan2 requires Zoltan2Core which requires Zoltan. It looks like your suggested fix is right. Are you concretizing with the original or clingo? I can cherry-pick your commit onto this branch.

Comment (balay) [reply var/spack/repos/builtin/packages/xsdk/package.py]: > Are you concretizing with the original or clingo

I have no idea what they are. I guess I'm using whatever spack defaults to..

> I can cherry-pick your commit onto this branch

Sure.

**Review by eugeneswalker — DISMISSED** [review]

**Comment by sethrj** [comment]
@eugeneswalker I'm merging in develop to trigger another gitlab ci pipeline to make sure everything still concretizes and builds with that. If it works, I'll need reapproval. I appreciate the review!

**Comment by eugeneswalker** [comment]
Pipeline failing because:

```
==> Error: trying to set variant "teuchos" in package "trilinos", but the package has no such variant [happened during concretization of trilinos@13.0.1%gcc@7.5.0+amesos+amesos2+anasazi+aztec+belos+epetra+epetraext+ifpack+ifpack2+intrepid+intrepid2+isorropia+kokkos+minitensor+ml+muelu+nox+phalanx+piro+rol+rythmos+sacado+shards+shylu+stk+stokhos+stratimikos+superlu-dist+teko+tempus+teuchos+tpetra+trilinoscouplings+zoltan+zoltan2 gotype=long arch=linux-ubuntu18.04-x86_64]
```

Can you add a commit here that removes `teuchos` from:
1. https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml#L287
2. https://github.com/spack/spack/blob/develop/share/spack/gitlab/cloud_pipelines/stacks/e4s-on-power/spack.yaml#L286

?

**Comment by sethrj** [comment]
Interesting, I thought that package specs ignored nonexistent variants -- I guess that's just when applying combinatorics rather than the main spec? Thanks and I'll fix.

**Comment by sethrj** [comment]
@eugeneswalker success! Can you please reapprove? 😄

**Review by eugeneswalker — APPROVED** [review]
Thank you!

File: var/spack/repos/builtin/packages/dealii/package.py [hunk 2]
```diff
@@ -210,16 +210,11 @@ class Dealii(CMakePackage, CudaPackage):
...
-               when='+trilinos+mpi+int64~cuda')
+    depends_on('trilinos+amesos+aztec+epetra+ifpack+ml+muelu+sacado', when='+trilinos')
+    depends_on('trilinos~hypre', when='+trilinos+int64')
     # TODO: temporary disable Tpetra when using CUDA due to
     # namespace "Kokkos::Impl" has no member "cuda_abort"
-    depends_on('trilinos@master+amesos+aztec+epetra+ifpack+ml+muelu+rol+sacado+teuchos~amesos2~ifpack2~intrepid2~kokkos~tpetra~zoltan2',
-               when='+trilinos+mpi~int64+cuda')
-    depends_on('trilinos@master+amesos+aztec+epetra+ifpack+ml+muelu+rol+sacado+teuchos~hypre~amesos2~ifpack2~intrepid2~kokkos~tpetra~zoltan2',
-               when='+trilinos+mpi+int64+cuda')
+    depends_on('trilinos@master+rol~amesos2~ifpack2~intrepid2~kokkos~tpetra~zoltan2', when='+trilinos+cuda')
```
Comment (jppelteret) [inline var/spack/repos/builtin/packages/dealii/package.py:217]: Thanks for the ping about these changes @sethrj. It looks much better now with all of these simplifications!

**Pull request merged**